src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2010 @@
+    val hi = mmin(until, length)
+    val newlen = hi - lo
+    if(newlen == length) this
+    else if(newlen <= 0) NVector0
+    else slice0(lo, hi)
+  }
+
+  /** Number of slices */
+  protected[immutable] def vectorSliceCount: Int
+  /** Slice at index */
+  protected[immutable] def vectorSlice(idx: Int): Array[_ <: AnyRef]
+  /** Dimension of the slice at index */
+  protected[immutable] final def vectorSliceDim(idx: Int): Int = {
+    val c = vectorSliceCount/2
+    c+1-abs(idx-c)
+  }
+  /** Length of all slices up to and including index */
+  protected[immutable] def vectorSlicePrefixLength(idx: Int): Int
+
+  override def iterator: Iterator[A] = new NVectorIterator(this)
+
+  override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int = iterator.copyToArray(xs, start, len)
+
+  //override def toVector: Vector[A] = this
+
+  //override protected def applyPreferredMaxLength: Int = Vector.defaultApplyPreferredMaxLength
+
+  override def stepper[S <: Stepper[_]](implicit shape: StepperShape[A, S]): S with EfficientSplit = {
+    import convert.impl._
+    val s = shape.shape match {
+      case StepperShape.IntShape    => new IntNVectorStepper(new NVectorIterator(this).asInstanceOf[NVectorIterator[Int]])
+      case StepperShape.LongShape   => new LongNVectorStepper(new NVectorIterator(this).asInstanceOf[NVectorIterator[Long]])
+      case StepperShape.DoubleShape => new DoubleNVectorStepper(new NVectorIterator(this).asInstanceOf[NVectorIterator[Double]])
+      case _                        => shape.parUnbox(new AnyNVectorStepper[A](new NVectorIterator(this)))
+    }
+    s.asInstanceOf[S with EfficientSplit]
+  }
+}
+
+
+/** Empty vector */
+private final object NVector0 extends NVector[Nothing] {
+  import NVectorStatics._
+
+  def length = 0
+  def apply(index: Int) = throw new IndexOutOfBoundsException
+
+  override def updated[B >: Nothing](index: Int, elem: B): NVector[B] = throw new IndexOutOfBoundsException
+
+  override def appended[B >: Nothing](elem: B): NVector[B] = new NVector1(wrap1(elem))
+
+  override def prepended[B >: Nothing](elem: B): NVector[B] = new NVector1(wrap1(elem))
+
+  override def iterator: Iterator[Nothing] = Iterator.empty
+
+  override def foreach[U](f: Nothing => U): Unit = ()
+
+  override def map[B](f: Nothing => B): NVector[B] = this
+
+  override def tail: NVector[Nothing] = throw new UnsupportedOperationException("empty.tail")
+
+  override def init: NVector[Nothing] = throw new UnsupportedOperationException("empty.init")
+
+  protected[this] def slice0(lo: Int, hi: Int): NVector[Nothing] = this
+
+  protected[immutable] def vectorSliceCount: Int = 0
+  protected[immutable] def vectorSlice(idx: Int): Array[_ <: AnyRef] = null
+  protected[immutable] def vectorSlicePrefixLength(idx: Int): Int = 0
+
+  override def equals(o: Any): Boolean = {
+    if(this eq o.asInstanceOf[AnyRef]) true
+    else o match {
+      case that: NVector[_] => false
+      case o => super.equals(o)
+    }
+  }
```

Comment:
We should keep optimized implementations of `equals` / `sameElements` for `Vector` / `Vector` comparisons short circuiting when the lengths differ and avoiding element comparison when `this` and `that` structurally share a backing array.

Reply:
The old optimizations don't make sense here. The only case where you would benefit from the shortcut is when comparing the result of `map(identity)`. If the vectors were built separately, they won't share any arrays. Doing structural comparisons on each array separately might be useful. This could cover cases such as dropping the head element and then consing it back on. The implementation would be rather complex and you need to take care not to compromise the more common cases though. You cannot handle all arrays individually because equal vectors can be balanced differently. In the old Vector the only possible difference was the offset but now we have to deal with all the different prefix lengths.

Reply:
@szeiger Would `(NVector.tabulate(1000)(i => i)).equals(NVector.tabulate(999)(i => i))` short circuit based on a cheap length call, [as per `Vector`](https://github.com/scala/scala/blob/802f8912f81958c52678fab275a86b7d0899c1c2/src/library/scala/collection/immutable/Vector.scala#L688)?

Reply:
Yes, this uses the default implementation of `sameElements` which checks `knownSize` first (and the size of an `IndexedSeq` is always known).